src/cdomains/pthreadDomain.ml

```diff
@@ -0,0 +1,96 @@
+open Prelude
+
+(** Thread ID *)
+module Tid = IntDomain.Flattened
+
+(** Context hash for function calls *)
+module Ctx = IntDomain.Flattened
+
+(** Set of predecessor nodes *)
+module Pred = struct
+  module Base = Basetype.ProgLocation
+  include SetDomain.Make (Base)
+
+  let of_node = singleton % MyCFG.getLoc
+
+  let of_current_node () = of_node @@ Option.get !MyCFG.current_node
+
+  let string_of_elt = Basetype.ProgLocation.show
+end
+
+type domain =
+  { tid : Tid.t
+  ; pred : Pred.t
+  ; ctx : Ctx.t
+  }
+[@@deriving to_yojson]
+
+module D = struct
+  include Printable.Std
+
+  type t = domain [@@deriving to_yojson]
+
+  (** printing *)
+  let show x =
+    Printf.sprintf
+      "{ Tid=%s; pred=%s; ctx=%s }"
+      (Tid.show x.tid)
+      (Pretty.sprint 200 (Pred.pretty () x.pred))
+      (Ctx.show x.ctx)
+
+
+  include Printable.PrintSimple (struct
+    type t' = t
+
+    let name () = "pthread state"
+
+    let show = show
+  end)
+
+  (** let equal = Util.equals *)
+  let equal x y =
+    Tid.equal x.tid y.tid && Pred.equal x.pred y.pred && Ctx.equal x.ctx y.ctx
+
+
+  (** compare all fields with correspoding compare operators *)
+  let compare x y =
+    List.fold_left
+      (fun acc v -> if acc = 0 && v <> 0 then v else acc)
+      0
+      [ Tid.compare x.tid y.tid
+      ; Pred.compare x.pred y.pred
+      ; Ctx.compare x.ctx y.ctx
+      ]
+
+
+  (** let hash = Hashtbl.hash *)
+  let hash x = Hashtbl.hash (Tid.hash x.tid, Pred.hash x.pred, Ctx.hash x.ctx)
+
+  let make tid pred ctx = { tid; pred; ctx }
+
+  let bot () = { tid = Tid.bot (); pred = Pred.bot (); ctx = Ctx.bot () }
+
+  let is_bot x = x = bot ()
```

Comment:
Should probably be like `is_top` instead of hoping that structural equality does the right thing.

Reply:
This is still to be done.